cluster-autoscaler/processors/nodeinfos/refine_node_infos_processor.go

```diff
@@ -0,0 +1,178 @@
+	RefineUsingSimilarNodeGroups bool
+	nodeInfoCache                nodeInfoCache
+}
+
+type nodeInfoCache struct {
+	sync.RWMutex
+	nodeInfos   map[string]*schedulerframework.NodeInfo
+	lastUpdated time.Time
+}
+
+// Process refines nodeInfos obtained by TemplateInfos when possible
+func (p *RefineNodeInfosProcessor) Process(ctx *context.AutoscalingContext, nodeInfosForNodeGroups map[string]*schedulerframework.NodeInfo) (map[string]*schedulerframework.NodeInfo, error) {
+	inaccurateNodeInfos := make(map[string]*schedulerframework.NodeInfo)
+	templatesNodeInfo := make(map[string]*schedulerframework.NodeInfo)
+
+	// for now this processor only supports refining by leveraging similar nodegroups
+	if !p.RefineUsingSimilarNodeGroups {
+		return nodeInfosForNodeGroups, nil
+	}
+
+	nodeGroups := make(map[string]cloudprovider.NodeGroup)
+	for _, nodeGroup := range ctx.CloudProvider.NodeGroups() {
+		nodeGroups[nodeGroup.Id()] = nodeGroup
+	}
+
+	if err := p.nodeInfoCache.update(nodeGroups); err != nil {
+		return nodeInfosForNodeGroups, err
+	}
+
+	ignoredTaints := make(taints.TaintKeySet)
+	for _, taintKey := range ctx.IgnoredTaints {
+		ignoredTaints[taintKey] = true
+	}
+
+	// build comparable (all from templates) NodeInfos
+	for groupID, nodeInfo := range nodeInfosForNodeGroups {
+		templatedNodeInfo, ok := p.nodeInfoCache.nodeInfos[groupID]
+		if !ok || templatedNodeInfo == nil {
+			continue
+		}
+
+		templatesNodeInfo[groupID] = templatedNodeInfo
+
+		if utils.IsNodeInfoBuiltFromTemplate(nodeInfo) {
+			inaccurateNodeInfos[groupID] = templatedNodeInfo
+		}
+	}
+
+	// refine inaccurate nodeinfos when we can find similar nodeInfos built from real nodes
+	for groupID, nodeInfo := range inaccurateNodeInfos {
+		nodeGroup, ok := nodeGroups[groupID]
+		if !ok {
+			continue
+		}
+
+		similars, err := p.NodeGroupSetProcessor.FindSimilarNodeGroups(ctx, nodeGroup, templatesNodeInfo)
+		if err != nil {
+			klog.Warningf("Failed to lookup for matching node groups for %s: %v", groupID, err)
+			return nodeInfosForNodeGroups, nil
+		}
+
+		for _, nodeGroup := range similars {
+			similarInfo, found := nodeInfosForNodeGroups[nodeGroup.Id()]
+			if !found || utils.IsNodeInfoBuiltFromTemplate(similarInfo) {
+				continue
+			}
+			refinedNodeInfo, err := mirrorNodeInfoFromSimilarGroup(nodeInfo, similarInfo, groupID, ignoredTaints)
+			if err != nil {
+				return nodeInfosForNodeGroups, err
+			}
+			nodeInfosForNodeGroups[groupID] = refinedNodeInfo
+			break
+		}
+	}
+
+	return nodeInfosForNodeGroups, nil
+}
+
+// CleanUp cleans up processor's internal structures.
+func (p *RefineNodeInfosProcessor) CleanUp() {
+}
+
+func mirrorNodeInfoFromSimilarGroup(oldInfo, newInfo *schedulerframework.NodeInfo, nodeGroupName string, ignoredTaints taints.TaintKeySet) (*schedulerframework.NodeInfo, errors.AutoscalerError) {
+	newInfoCopy := newInfo.Clone()
+
+	// keep locality labels (region, zone, failure domain, ...) from original template
```

Comment:
Is this really all we need from template? I think it would probably be good to have things like node name be different between templates. Another example are daemonset pods (that I think will be part of nodeInfo) - how do we know there are no zonal DS? Not sure about other stuff, but it would be good to double check if there is anything else worth randomizing or keeping from old template.

Reply:
That's a very good call. Changed for a conservative approach: start from (clone) the original/template NodeInfo, and only pick the node from the model/real world NodeInfo (keeping the original node's name, uid, hostname, locality labels)